src/initializers.jl

```diff
@@ -26,6 +26,7 @@ const SA_F64 = SA{Float64}
 @inline similar_type(::Type{SA}, ::Size{S}) where {S} = SArray{Tuple{S...}}
 @inline similar_type(::Type{SA{T}}, ::Size{S}) where {T,S} = SArray{Tuple{S...}, T}
 
+# Lower bound `SA{E}<:T` prevents `T == Union{}` from accidentally matching this signature.
 @inline Base.getindex(sa::Type{T}, xs...) where {E, SA{E}<:T<:SA} = similar_type(sa, Size(length(xs)))(xs)
```

Comment:
What about `SA` precisely?

Reply:
There's nothing special about `SA`... It's just that `::Type{<:SA}` is considered more specific to matching `Union{}` than a generic `::Type{T}`:

```julia
julia> foo(::Type{T}) where {T} = "generic"
julia> foo(::Type{<:Int}) = "special"
julia> foo(Union{})
"special"
```

Reply:
somewhat disturbing eh?

Reply:
Yes, I get that part.

I can see the above is fine for `SA{Int64}[1,2,3]`. But what about vanilla `SA[1,2,3]`? We don't have `SA{E} <: SA` for any (invariant) `E`, do we?

Reply:
Oh right, I didn't understand your question. You're asking why `T === SA` works here?

We do have `SA{E} <: SA` for any choice of `E`, but then `E` doesn't have a definite value which is kind of weird. Actually this kind of thing has caused the inliner to fail in the past (see #665) for some reason.

The normalization of the following is also doing my head in a bit

```
julia> T where SA{E} <: T <: SA where E
SA
```